### DISCUSSION TITLE
Not able to launch appium espresso driver

### DISCUSSION THREAD

**Discussion by kianreno** [appium/appium#17898] [Q&A]
I am getting an error when I try launch the espresso server. It successfully builds and then gives a warning that it has crashed without giving much detail. If I completely remove the additionalAppDependencies section of the espresso build config then I can get the server to launch but then I can't navigate in the app at all. I am new to using the espresso driver and am probably not setting it up properly but was hoping someone could help. 

Desired Capabilities (I kept throwing in extra dependencies into the espresso build config to see if it would help): 

[android_caps_compose.txt](https://github.com/appium/appium/files/10190478/android_caps_compose.txt)

`{
  "appium:noReset": "true",
  "appium:settings[driver]": "compose",
  "appium:forceEspressoRebuild": "true",
  "appium:newCommandTimeout": "300",
  "remoteWebDriverType": "Appium",
  "appium:platformName": "Android",
  "remoteMobileDriver": "ANDROID_DRIVER",
  "remoteWebDriverUrl": "http://localhost:4723/",
  "appium:deviceName": "emulator-5554",
  "appium:espressoBuildConfig": "{  \"toolsVersions\" : {\"compileSdk\" : 32,
  \"jvmTarget\" : \"11\",
  \"composeVersion\" : \"1.2.0-rc03\", \"annotationVersion\" : \"1.1.0\"
}, \"additionalAndroidTestDependencies\" : [ \"androidx.hilt:hilt-navigation-compose:1.0.0\", \"androidx.compose.ui:ui:1.2.0-rc03\", \"androidx.compose.ui:ui-test-junit4:1.2.0-rc03\", \"androidx.compose.runtime:runtime-rxjava2:1.2.0-rc03\", \"androidx.navigation:navigation-compose:2.4.2\", \"androidx.test:core:1.4.0\", \"org.jetbrains.kotlinx:kotlinx-coroutines-test:1.6.3\", \"androidx.constraintlayout:constraintlayout:2.1.4\", \"androidx.test:core-ktx:1.4.0\", \"androidx.compose.material:material:1.2.0-rc03\", \"androidx.test:runner:1.4.0\", \"androidx.test.ext:junit:1.1.3\",  \"androidx.lifecycle:lifecycle-viewmodel-compose:2.5.0\" ] ,
    \"additionalAppDependencies\" : [ \"androidx.compose.material:material:1.2.0-rc03\",  \"androidx.lifecycle:lifecycle-viewmodel-compose:2.5.0\", \"androidx.compose.ui:ui:1.2.0-rc03\", \"androidx.compose.ui:ui-test-junit4:1.2.0-rc03\", \"androidx.test:runner:1.4.0\", \"androidx.test:core:1.4.0\", \"androidx.test.ext:junit:1.1.3\",  \"org.jetbrains.kotlinx:kotlinx-coroutines-test:1.6.3\" ]
}",
  "appium:automationName": "espresso",
  "appium:showGradleLog": "true"
}
`

Details: 
Mac OS 12.6
Appium v2.0.0-beta46
Espresso driver v2.13.8
device: emulated pixel 5 with api 32
App under test is written with jetpack compose
Node v19.2.0
npm v8.19.3

appium command line output: 

[appium_output.log](https://github.com/appium/appium/files/10190463/appium_output.log)

**Comment by portyu9** [comment]
Hi @kianreno 

The problem is most likely the `espressoBuildConfig`, not Compose itself.

You are injecting a large number of Compose/AndroidX dependencies into both:

```text
additionalAppDependencies
additionalAndroidTestDependencies
```

That can create runtime dependency conflicts and cause the Espresso server to crash after it successfully builds.

Start with a minimal configuration and make the Espresso server versions match the AUT exactly:

```json
{
  "toolsVersions": {
    "compileSdk": 32,
    "composeVersion": "1.2.0-rc03"
  }
}
```

Keep:

```json
"appium:forceEspressoRebuild": true,
"appium:settings[driver]": "compose"
```

Do **not** add every application dependency manually.

Only add a dependency when `logcat` reports a specific missing class/method, and use the **exact version used by the application**. Libraries such as `ui-test-junit4`, `androidx.test:runner`, and `androidx.test.ext:junit` belong on the Android-test side rather than being duplicated as application dependencies.

With current Espresso Driver, also run:

```bash
appium driver run espresso diagnose-app --app /path/to/app.apk
```

It checks Compose/AndroidX dependency alignment and other conditions that commonly cause Espresso server startup failures.

So the fix is: **remove the duplicated dependency list, align Compose/Kotlin/AndroidX versions with the AUT, force one clean Espresso rebuild, and add dependencies back only when the actual crash log proves they are required.**

Please mark this answer as accepted if it helped, thank you.
